### PR TITLE
Added Standard Resolution Dropdown List in Settings

### PR DESCRIPTION
**Pull request by newyork167**
Adds a dropdown to the settings screen with a list of common/standard resolutions in differing aspect ratios.

### CODE REVIEW

**Comment by matjaeck** [comment]
I'm wondering if we want a general setting for resolution and then just say open it in full screen or windowed.

**Comment by pchote** [comment]
Full-screen resolution controls depend on #10382, which still needs some serious work to complete.

**Comment by newyork167** [comment]
For the logicArgs in the settings yaml, how do you guys feel about this kind of setup?
image

**Comment by matjaeck** [comment]
Please paste code snippets by using the formatting explained at https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code.

**Comment by pchote** [comment]
If you go with a simpler yaml syntax:
```yaml
SupportedResolutions:
    1024, 768: 1024x768 - 4:3
    1920, 1080: 1920x1080 - 16:9
    ...
```

then you can use the following snippet to automatically parse it into a dictionary:
```csharp
MiniYaml yaml;
if (logicArgs.TryGetValue("SupportedResolutions", out yaml))
{
	var resolutions = (Dictionary<int2, string>)FieldLoader.GetValue("SupportedResolutions", typeof(Dictionary<int2, string>), yaml, null);
	// Do things with resolutions
}
```
You can invert the `int2` and `string` if you want to keep the same ordering as in the current code, but IMO having the int2 on the key makes more sense.

Note that our minimum supported resolution is 1024x720, so you should remove all the options smaller than this.

**Comment by pchote** [comment]
Any progress here @newyork167?

**Comment by newyork167** [comment]
> Any progress here @newyork167?

Sorry @pchote work has kept me busy the last few weeks. I've pushed some changes reflecting your new settings format, let me know what you think.

**Comment by Mailaender** [comment]
You could also use call `SDL_GetDisplayMode()` in a loop, `SDL_GetNumDisplayModes()` times according to https://wiki.libsdl.org/MigrationGuide

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 1]
```diff
@@ -171,6 +172,15 @@ Action InitDisplayPanel(Widget panel)
 			windowModeDropdown.GetText = () => ds.Mode == WindowMode.Windowed ?
 				"Windowed" : ds.Mode == WindowMode.Fullscreen ? "Fullscreen (Legacy)" : "Fullscreen";
 
+			// Show Custom Resolution Input or Most Common Dropdown
+			var useCustomResolutionCheckbox = panel.Get<CheckboxWidget>("USE_CUSTOM_RESOLUTION_CHECKBOX");
+			var useCustomResolutionOnClick = useCustomResolutionCheckbox.OnClick;
+			useCustomResolutionCheckbox.IsVisible = () => ds.Mode == WindowMode.Windowed;
+			useCustomResolutionCheckbox.OnClick = () =>
+			{
+				useCustomResolutionOnClick();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: I don't think this is needed when you don't perform any extra options. (I.e. It should be save to remove line 177, and lines 179-182.)

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 2]
```diff
@@ -619,6 +640,33 @@ static bool ShowWindowModeDropdown(DropDownButtonWidget dropdown, GraphicSetting
...
 
+		static bool ShowDisplayResolutionDropdown(DropDownButtonWidget dropdown, GraphicSettings s, Dictionary<string, MiniYaml> logicArgs)
+		{
+			var options = new Dictionary<string, int2>();
+			MiniYaml yaml;
+			if (logicArgs.TryGetValue("SupportedResolutions", out yaml))
+			{
+				var resolutions = (Dictionary<int2, string>)FieldLoader.GetValue("SupportedResolutions", typeof(Dictionary<int2, string>), yaml, null);
+				foreach (var resolution in resolutions)
+				{
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Please drop the braces around this single-line statement.

**Review by abcdefg30 — CHANGES_REQUESTED** [review]
Sorry for the delayed review. A few things:
- You will also need to adjust `cnc`'s `settings.yaml` chrome file.
- Is it intentional, that the dropdown shows the current custom resolution when you uncheck the checkbox? (That way you can get entries that are not in the list.)
- Have you looked at Mailaender's suggestion on how to generate the list? Is that something we want to do? (I think it might be worth a shot.)

Looks good to me otherwise.

**Comment by reaperrr** [comment]
There hasn't been any progress in quite a while, so I'll close this to clean up the PR queue and move it to Future milestone. Hopefully someone will pick this up.

**Pull request closed without merging**